audiobookshelf: init at 2.2.15

jvanbruegge opened the pull request:
###### Description of changes

Audiobookshelf is a self-hosted alternative to services like audible

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

jvanbruegge committed:
tone: init at 0.1.3

jvanbruegge committed:
audiobookshelf: init at 2.2.15

jvanbruegge commented:
I don't understand the error that ofborg throws

jvanbruegge commented:
I used `buildNpmPackage` for the client now so I don't have to commit a 14k LOC file with the dependencies

jvanbruegge commented:
But I still don't know what ofborg wants here

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/access-to-absolute-path-is-forbidden-in-restricted-mode/25058/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/access-to-absolute-path-is-forbidden-in-restricted-mode-ofborg/25058/2

SuperSandro2000 commented on pkgs/servers/audiobookshelf/default.nix:
```diff
@@ -0,0 +1,120 @@
+  '';
+
+  passthru.updateScript = ./generate.sh;
+
+
+  meta = with lib; {
```
```suggestion

  meta = with lib; {
```

SuperSandro2000 commented on pkgs/servers/audiobookshelf/default.nix:
```diff
@@ -0,0 +1,120 @@
+    cp -r ${client}/lib/node_modules/${pname}-client/dist $out/opt/client/dist
+    mkdir $out/bin
+
+    echo '#!${stdenv.shell}
```
We should put the wrapper into an extra file if it gets that big.

SuperSandro2000 commented on pkgs/servers/audiobookshelf/default.nix:
```diff
@@ -0,0 +1,120 @@
+in stdenv.mkDerivation rec {
+  inherit pname version src;
+
+  buildInputs = [ nodejs ffmpeg-full tone util-linux ];
```
ffmpeg-full and tone are only used in the wrapper, so putting them here will be a noop.

SuperSandro2000 commented on pkgs/servers/audiobookshelf/default.nix:
```diff
@@ -0,0 +1,120 @@
+    pname = "${pname}-client";
+    inherit version;
+
+    src = runCommand "cd" {} ''
```
```suggestion
    src = runCommand "cp-source" {} ''
```

SuperSandro2000 commented on pkgs/servers/audiobookshelf/generate.sh:
```diff
@@ -0,0 +1,17 @@
+cd "$(dirname "${BASH_SOURCE[0]}")" || exit 1
+
+hash=$(prefetch-npm-deps "$src/client/package-lock.json")
+echo "npmDepsHash = \"$hash\";" > test.txt
```
Is that file even used?

jvanbruegge replied:
Nope, that was left over from debugging the sed command

SuperSandro2000 commented on pkgs/applications/audio/tone/default.nix:
```diff
@@ -0,0 +1,38 @@
+
+in buildDotnetModule rec {
+  pname = "tone";
+  baseName = pname;
```
You are sure baseName is required?

jvanbruegge replied:
Copied it from another dotnet derivation, I don't think so

SuperSandro2000 commented on pkgs/servers/audiobookshelf/default.nix:
```diff
@@ -0,0 +1,120 @@
+    npmDepsHash = "sha256-UFmdTWxkOGO73g45NzB51/d1oGOhfNZtOlKSfbTsxPY=";
+  };
+
+  nodeDeps = (import ./node-composition.nix {
+    inherit pkgs nodejs;
+    inherit (stdenv.hostPlatform) system;
+  }).nodeDependencies;
```
I think this line is causing IFD which is forbidden. I think you must commit the package.json.

jvanbruegge replied:
I found another solution in the derivation for jellyfin, let's see if that works

jvanbruegge replied:
Looks like it does

jvanbruegge commented:
Nice, the tests pass now

dit7ya commented:
Hey @jvanbruegge, have you seen my PR at #201552? I want to update one little thing in the nixos module, but other than that I was happy with. It does not generate node-deps and has a much shorter diff.

jvanbruegge commented:
No I did not see that, which is weird because I did use the github search to see if it was already being packaged.

I've done the same as you now for the package, even though that means that it cannot have an update-script any more. (There is no option that simulates --only-production in prefetch-npm-deps).

I also updated audiobookshelf to the newest version (2.2.14).

Could you retarget your branch to this PR and add the service and tests in a seperate commit? I want to keep the wrapper from this PR because I want to run it without the nixos module and without a config file later on.

@SuperSandro2000 I have addressed all your comments

jvanbruegge commented:
@SuperSandro2000 I've updated to the newest release

advplyr commented:
Hey, heads up the next release coming in a few days is migrating the file-based JSON database to a sqlite database. This branch https://github.com/advplyr/audiobookshelf/tree/sqlite_2 has the working code. 

This is using [node-sqlite3](https://github.com/TryGhost/node-sqlite3). I have tested the Docker images and debian package. I'm happy to discuss this more if you want here or in discord/matrix.
